h5py/h5a.pyx

```diff
@@ -47,6 +47,9 @@ def create(ObjectID loc not None, char* name, TypeID tid not None,
         Link access property list for obj_name
     """
 
+    if space.get_simple_extent_npoints() > 3116:
+        # reset the dataspace extent
+        space.set_extent_simple(())
```

Comment:
I don't fully understand this, and it now seems completely separate from the get/set_attr_phase_change methods. The changes to expose those two HDF5 methods in the low-level API look fine - would you like to do those changes by themselves first, and we can merge them pretty quickly, and then discuss this in a separate PR?

If you want to keep them together, can you explain what this is meant to do? I don't think modifying the extent (shape) of the data space makes sense. And what is the significance of 3116?

Reply:
I just want to try to modify the data space of an attribute to solve the problem (at least test cases pass). This is really different from the previous changes. I will still try to use set/get_attr_phase_change, but I still have some questions, as I comment before: 
> the problem is we create the group/dataset before creating attributes, and in many cases, we don't even create attributes, we can't know the value info about the attributes before we create it.

In my understanding, we should set `max_compact&min_dense` parameter in set_attr_phase_change() according the size of an attribute when creating a group/dataset, that is to say, before we set the group/dataset, we need to know the size of the attribute exactly. But the reality is that we don't know.

so I don't know how to made it. the problem in https://github.com/h5py/h5py/issues/1053#issue-330325080 is that object header message space is limited, so I realize we can change attribute's data space to break its limitation. but I know this is away from our original plan.
The critial value is from #1053, I don't know why it's 3117, but not 65535:

> We generally consider the maximum size of an attribute to be 64K bytes. The library has two ways of storing attributes larger than 64K bytes: in dense attribute storage or in a separate dataset. Using dense attribute storage is described in this section, and storing in a separate dataset is described in the next section.

so there are two ways:
a) set `set_attr_phase_change(0,0)` in the right place to store an attribute in dense (this will be very similar to store large number of attributes).
b) create a new dataset to store that large attribute.

@takluyver what do you think?

Reply:
As far as I know, there isn't anything we can do to *automatically* solve #1053. What you're doing here makes the error disappear, but it doesn't store all the data. I've just tried adding a line to your test to check the value of the attribute:

```python
np.testing.assert_array_equal(grp.attrs['data'], data)
```

And only the first value from the array is stored. This is worse than the error!

Exposing the attr_phase_change methods means someone who knows they need to store larger attributes can do so, even if requires a few lines of extra code. That's a useful step by itself.

Reply:
Do you mean, at the moment, we just exposing these two methods? if so, then I will remove the related tests in test_attrs.py

P.S. in my local, if I turn on the track_order when I creating file/group/dataset, then these test cases will pass all. 
but if I test with the code, but outsite of the h5py project, there will occur `record is already in B-tree` error.